pkg/controller/nstemplateset/nstemplateset_controller.go

```diff
@@ -286,15 +312,56 @@ func (r *NSTemplateSetReconciler) ensureInnerNamespaceResources(logger logr.Logg
 	return nil
 }
 
-// nextNamespaceToProvision returns first namespace (from given namespaces) with
-// namespace status is active and revision not set
-// or namespace present in tcNamespaces but not found in given namespaces
-func nextNamespaceToProvision(tcNamespaces []toolchainv1alpha1.NSTemplateSetNamespace, namespaces []corev1.Namespace) (*toolchainv1alpha1.NSTemplateSetNamespace, *corev1.Namespace, bool) {
-	for _, tcNamespace := range tcNamespaces {
+func (r *NSTemplateSetReconciler) deleteRedundantObjects(nsType string, params map[string]string, namespace *corev1.Namespace, newObjects []runtime.RawExtension) error {
+	currentTier := namespace.Labels[toolchainv1alpha1.TierLabelKey]
+	currentTmpl, err := r.getTemplateContent(currentTier, nsType)
+	if err != nil {
+		return errs.Wrapf(err, "failed to to retrieve template of the current tier '%s' for namespace '%s'", currentTier, namespace.Name)
+	}
+
+	tmplProcessor := template.NewProcessor(r.client, r.scheme)
+	currentObjs, err := tmplProcessor.Process(currentTmpl, params, template.RetainAllButNamespaces)
```

Comment:
This will work for the usecase when we change the tier. But I wonder if we could just load all the objects in the namespace with `labels["provider"] == "codeready-toolchain"`, so, we get the actual objects and not just objects which are defined in the used tier. In this case we could re-use it later for template updates. WDYT?
But I'm fine with keeping it as is for now and address the template update case separately later.

Reply:
Good idea, but I'm not sure if there is some generic list-object that we could use for that operation. I'm afraid that we would need to know which kinds of objects we want to list and list them separately. I tried to use the `UnstructuredList` to list all objects but it failed exactly with the error that I expected: `Object 'Kind' is missing in 'unstructured object has no kind'`.
We cannot make the "list request" because without knowing the GVK we cannot contsruct the API url for that.

Reply:
I know that this will work only for changes of the tier; for the updates of the templates itself we have two options:
* store all revisions of the tiers in OS
* download the older revision from GH

I would go with the second option - downloading it from GH

Reply:
Yeah... I was thinking about using the dynamic client but we would still need to know the GVK to list objects... We could get all GVKs available in the API Server but it would be very time/resource consuming to get the list of all GVKs and then list the corresponding resources in the namespace.

> - store all revisions of the tiers in OS
> - download the older revision from GH

I would avoid depending on GitHub in runtime. Storing all the revisions in etcd should be pretty easy to implement and use.